File: src/main/java/org/pragmatica/example/ticketing/shared/PriceTier.java
package org.pragmatica.example.ticketing.shared;

import java.util.Locale;

import org.pragmatica.aether.slice.mapping.ValueMapping;
import org.pragmatica.lang.Cause;
import org.pragmatica.lang.Result;


public enum PriceTier {
    PREMIUM,
    STANDARD,
    ECONOMY,
    ACCESSIBLE,
    RESTRICTED_VIEW;
    public sealed interface Error extends Cause {
        record Unknown(String raw) implements Error {
            @Override
            public String message() {
                return "Unknown price tier: " + raw;
            }
        }

        static Error unknown(String raw) {
            return new Unknown(raw);
        }
    }
    public static Result<PriceTier> priceTier(String raw) {
        return Result.lift(Error.unknown(raw),
                           () -> PriceTier.valueOf(raw.trim().toUpperCase(Locale.ROOT)));
    }
    /// Store-boundary descriptor: `pg-codegen` lowers a `PriceTier` column to its uppercase enum name
    /// and lifts a raw `tier` column back through `priceTier(...)`, so a corrupt tier value fails the
    /// row decode instead of yielding an invalid tier.
    public static ValueMapping<PriceTier, String> valueMapping() {
        return ValueMapping.of(PriceTier::name, PriceTier::priceTier);
    }
}

File: fsharp/src/Ticketing/Shared/PriceTier.fs
namespace Ticketing.Shared

open Pragmatica

[<RequireQualifiedAccess>]
type PriceTierError =
    | Unknown of raw: string

    interface Cause with
        member this.Message =
            match this with
            | PriceTierError.Unknown raw -> $"Unknown price tier: {raw}"

[<RequireQualifiedAccess>]
type PriceTier =
    | Premium
    | Standard
    | Economy
    | Accessible
    | RestrictedView

[<RequireQualifiedAccess>]
module PriceTier =
    let parse (raw: string) : Result<PriceTier, Cause> =
        match raw.Trim().ToUpperInvariant() with
        | "PREMIUM" -> Ok PriceTier.Premium
        | "STANDARD" -> Ok PriceTier.Standard
        | "ECONOMY" -> Ok PriceTier.Economy
        | "ACCESSIBLE" -> Ok PriceTier.Accessible
        | "RESTRICTED_VIEW" -> Ok PriceTier.RestrictedView
        | _ -> Error(PriceTierError.Unknown raw :> Cause)

    /// Persistence/wire form: the Java enum constant name (`RESTRICTED_VIEW`), matching the `tier`
    /// column literals.
    let name (tier: PriceTier) =
        match tier with
        | PriceTier.Premium -> "PREMIUM"
        | PriceTier.Standard -> "STANDARD"
        | PriceTier.Economy -> "ECONOMY"
        | PriceTier.Accessible -> "ACCESSIBLE"
        | PriceTier.RestrictedView -> "RESTRICTED_VIEW"
